arangod/Aql/OptimizerRules.cpp

```diff
@@ -1719,13 +1720,28 @@ void arangodb::aql::moveCalculationsUpRule(Optimizer* opt,
   VarSet vars;
 
   for (auto const& n : nodes) {
-    auto nn = ExecutionNode::castTo<CalculationNode*>(n);
-
-    if (!nn->expression()->isDeterministic()) {
+    bool isAccessCollection = false;
+    if (!n->isDeterministic()) {
       // we will only move expressions up that cannot throw and that are
       // deterministic
+
       continue;
     }
+    if (n->getType() == EN::CALCULATION) {
+      auto nn = ExecutionNode::castTo<CalculationNode*>(n);
+      if (::accessesCollectionVariable(plan.get(), nn, vars)) {
+        isAccessCollection = true;
+      }
+    } else {
+      // if it's a subquery node, it cannot move upwards if there's a
+      // modification keyword in the subquery e.g.
+      // INSERT would not be scope-limited by the outermost subqueries, so we could end up
+      // inserting a smaller amount of documents than what's actually proposed in the query.
+      auto nn = ExecutionNode::castTo<SubqueryNode*>(n);
+      if (nn->isModificationNode()) { 
+        continue;
+      }
```

Comment:
Is this not already covered by the previous `isDeterministic` check?

Reply:
Yes! It won't do any harm leaving the check in here, but it is redundant. 
As it is a nice-to-have cleanup, shall we clean it up in all backports, or only do it in devel? I would prefer the latter.

Reply:
I am ok with only doing it in devel.

Reply:
Thanks!